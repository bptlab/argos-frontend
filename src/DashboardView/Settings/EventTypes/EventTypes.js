import React, { Component } from 'react';
import {argosConfig} from './../../../config/argosConfig';
import ConfirmOverlay from "../../../Utils/ConfirmOverlay/ConfirmOverlay";

class EventTypes extends Component {

    constructor(props) {
        super(props);
        this.state = {
            error: {
                errorCode: null,
                errorMessage: null
            },
            eventTypes: []
        };
        //function binding
        this.receiveEventTypes = this.receiveEventTypes.bind(this);
        this.receiveError = this.receiveError.bind(this);
        this.fetchEventTypes = this.fetchEventTypes.bind(this);
        this.handleDeleteEventType = this.handleDeleteEventType.bind(this);
        this.deletionSuccessful = this.deletionSuccessful.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        if(nextProps.settingsVisible) {
            this.fetchEventTypes();
            nextProps.dataSource.notificationService.subscribe("EventType", this.fetchEventTypes.bind(this));
        } else if(!nextProps.settingsVisible) {
            this.props.dataSource.notificationService.unsubscribe("EventType", this.fetchEventTypes.bind(this));
        }
    }
    
    receiveEventTypes(eventTypes) {
        this.setState({
            eventTypes: eventTypes
        });
    }
    
    receiveError(errorMessage, errorCode) {
        this.setState({
            error: {
                errorCode: errorCode,
                errorMessage: errorMessage
            }
        });
    }
    
    fetchEventTypes() {
        this.props.dataSource.fetchAllEventTypes(this.receiveEventTypes, this.receiveError);
    }
    
    deletionSuccessful() {
        this.fetchEventTypes();
    }

    handleDeleteEventType(eventType) {
        $("#delete-prompt-modal").modal('hide');
        this.props.dataSender.deleteEventType(eventType, this.deletionSuccessful, this.receiveError);
    }

    /* istanbul ignore next */
    handleAbortDelete() {
        $("#delete-prompt-modal").modal('hide');
    }
    
    matchError() {
        const errorCode = this.state.error.errorCode;
        const errorMessage = this.state.error.errorMessage;
        if(errorCode) {
            switch(errorCode) {
                case 403: {
                    return argosConfig.errorMessage403;
                }
                case 500: {
                    return argosConfig.errorRemainingDependencies + "<br/>Dependent eventTypes: " + errorMessage;
                }
                default:
                    return argosConfig.RESTInterfaceConnectionError;
            }
        } else {
            return "";
        }
    }

    render() {
        const errorMessage = this.matchError();
        const lineElements = [];
        this.state.eventTypes.forEach((eventType) => {
            const deleteCallback = this.handleDeleteEventType.bind(this, eventType);
            const abortCallback = this.handleAbortDelete.bind(this);
            lineElements.push(
                <tr key={eventType.id}>
                    <td>{eventType.id}</td>
                    <td>{eventType.name}</td>
                    <td>{eventType.eventQuery}</td>
                    <td className="center">
                        <a className="nav-link" data-toggle="modal" data-target="#delete-prompt-modal">
                            <i className="fa fa-trash" />
                        </a>
                        <ConfirmOverlay title="Delete Event Type?" onAbort={abortCallback} onSubmit={deleteCallback}
                                        id="delete-prompt" abortButtonText="Abort" submitButtonText="Submit">
                            {argosConfig.confirmEventTypeDeletion}
                        </ConfirmOverlay>
                    </td>
                </tr>
            );
        });
        
        return (
            <div className="event-types-settings">
                {errorMessage && 
                    (<div className="alert alert-danger" role="alert">
                        {errorMessage}
                    </div>)
                }
                <table className="table table-striped event-types-table">
                    <thead>
                        <tr>
                            <th>#</th>
                            <th>Name</th>
                            <th>Query</th>
                            <th className="center">
                                {argosConfig.tableHeaderDescriptionDelete}
                            </th>
                        </tr>
                    </thead>
                    <tbody>
                        {lineElements}
                    </tbody>
                </table>
                <p>{argosConfig.eventTypeAddHint}</p>
            </div>
        );
    }
}
export default EventTypes;