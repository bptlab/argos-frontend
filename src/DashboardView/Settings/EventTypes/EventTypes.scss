.event-types-table th, .event-types-table td {
  text-align: center;
}

.event-types-table .fa {
  cursor: pointer;
}