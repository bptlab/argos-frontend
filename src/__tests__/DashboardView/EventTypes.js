import React from 'react';
import ReactDOM from 'react-dom';
import EventTypes from './../../DashboardView/Settings/EventTypes/EventTypes.js';
import renderer from 'react-test-renderer';
import TestData from '../testData/frontend_eventTypes.js';
let instance, dataSource, dataSender, component;
import 'bootstrap';
import {argosConfig} from './../../config/argosConfig.js'

beforeEach(() => {
    const notificationService = {
        subscribe: jest.fn(),
        unsubscribe: jest.fn()
    };
    dataSource = {
        fetchAllEventTypes: jest.fn(),
        notificationService: notificationService
    };
    dataSender = {
        notificationService: notificationService,
        deleteEventType: jest.fn(),
    };
    component = renderer.create(
        <EventTypes
            dataSource={dataSource}
            dataSender={dataSender}
            settingsVisible={false}
            ref={(child) => {instance = child}} />
    );
});

test('Correct rendering of EventTypes-Settings', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Receive eventTypes', () => {
    instance.receiveEventTypes(TestData.EVENTTYPES);
    expect(instance.state.eventTypes).toEqual(TestData.EVENTTYPES);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Correct rendering of EventTypes-Settings while beeing visible', () => {
    const node = document.createElement('div');
    //two times rendering using ReactDOM invokes updating of component
    ReactDOM.render( <EventTypes
        dataSource={dataSource}
        dataSender={dataSender}
        settingsVisible={false}
        ref={(child) => {instance = child}} />, node);
    ReactDOM.render( <EventTypes
        dataSource={dataSource}
        dataSender={dataSender}
        settingsVisible={true}
        ref={(child) => {instance = child}} />, node);
    expect(dataSource.fetchAllEventTypes).toHaveBeenCalled();
});

test('Receiving 403 error', () => {
    const errorCode = 403;
    const errorMessage = "You are not allowed to touch this eventType";
    instance.receiveError(errorMessage, errorCode);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(instance.state.error.errorCode).toEqual(errorCode);
    expect(instance.state.error.errorMessage).toEqual(errorMessage);
});

test('Receiving 500 error', () => {
    const errorCode = 500;
    const errorMessage = "There still exist dependencies concerning this eventType";
    instance.receiveError(errorMessage, errorCode);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(instance.state.error.errorCode).toEqual(errorCode);
    expect(instance.state.error.errorMessage).toEqual(errorMessage);
});

test('Receiving undefined error', () => {
    instance.receiveError("Error", 999);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('Delete eventTypes', () => {
    instance.deleteEventType(TestData.EVENTTYPES);
    expect(dataSender.deleteEventType).toHaveBeenCalled();
});

test('Successfuly removed eventType', () => {
    instance.deletionSuccessful();
    expect(dataSource.fetchAllEventTypes).toHaveBeenCalled();
});