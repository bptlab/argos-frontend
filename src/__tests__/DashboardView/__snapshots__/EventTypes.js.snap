exports[`test Correct rendering of EventTypes-Settings 1`] = `
<div
  className="event-types-settings">
  
  <table
    className="table table-striped event-types-table">
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Name
        </th>
        <th>
          Query
        </th>
        <th
          className="center">
          Delete?
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <p>
    To add a new Event-Type please refer to a Product and press the \"Plus\"-Button.
  </p>
</div>
`;

exports[`test Receive eventTypes 1`] = `
<div
  className="event-types-settings">
  
  <table
    className="table table-striped event-types-table">
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Name
        </th>
        <th>
          Query
        </th>
        <th
          className="center">
          Delete?
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          0
        </td>
        <td>
          Error Events
        </td>
        <td />
        <td
          className="center">
          <a
            className="nav-link"
            data-target="#delete-prompt-modal"
            data-toggle="modal">
            <i
              className="fa fa-trash" />
          </a>
          <div
            className="delete-prompt modal fade"
            id="delete-prompt-modal"
            role="dialog"
            tabIndex="-1">
            <div
              className="modal-dialog modal-md"
              role="document">
              <div
                className="modal-content">
                <div
                  className="modal-header">
                  <h5
                    className="modal-title"
                    id="delete-prompt-label">
                    Delete Event Type?
                  </h5>
                </div>
                <div
                  className="modal-body">
                  Do you really want to delete this event type? This is a destructive action and cannot be undone.
                </div>
                <div
                  className="modal-footer">
                  <button
                    className="btn"
                    onClick={[Function]}
                    type="button">
                    <div>
                      Abort
                    </div>
                  </button>
                  <button
                    className="btn"
                    onClick={[Function]}
                    type="button">
                    <div>
                      Submit
                    </div>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <p>
    To add a new Event-Type please refer to a Product and press the \"Plus\"-Button.
  </p>
</div>
`;

exports[`test Receiving 403 error 1`] = `
<div
  className="event-types-settings">
  <div
    className="alert alert-danger"
    role="alert">
    Your are not allowed to modify this Event-Type
  </div>
  <table
    className="table table-striped event-types-table">
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Name
        </th>
        <th>
          Query
        </th>
        <th
          className="center">
          Delete?
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <p>
    To add a new Event-Type please refer to a Product and press the \"Plus\"-Button.
  </p>
</div>
`;

exports[`test Receiving 500 error 1`] = `
<div
  className="event-types-settings">
  <div
    className="alert alert-danger"
    role="alert">
    You can not delete this Event-Type, because of existing dependencies<br/>Dependent eventTypes: There still exist dependencies concerning this eventType
  </div>
  <table
    className="table table-striped event-types-table">
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Name
        </th>
        <th>
          Query
        </th>
        <th
          className="center">
          Delete?
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <p>
    To add a new Event-Type please refer to a Product and press the \"Plus\"-Button.
  </p>
</div>
`;

exports[`test Receiving undefined error 1`] = `
<div
  className="event-types-settings">
  <div
    className="alert alert-danger"
    role="alert">
    A critical connection error occurred!
  </div>
  <table
    className="table table-striped event-types-table">
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Name
        </th>
        <th>
          Query
        </th>
        <th
          className="center">
          Delete?
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <p>
    To add a new Event-Type please refer to a Product and press the \"Plus\"-Button.
  </p>
</div>
`;
